llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -1033,21 +1086,92 @@ splitDebugInfoSectionData(MutableArrayRef<char> SectionData,
   return SplitData;
 }
 
-Error MCCASBuilder::createDebugInfoSection(ArrayRef<DebugInfoCURef> CURefs) {
+Error MCCASBuilder::createDebugInfoSection(
+    ArrayRef<DebugInfoCURef> CURefs, DebugAbbrevOffsetsRef AbbrevOffsetsRef) {
   if (CURefs.empty())
     return Error::success();
 
   startSection(DwarfSections.DebugInfo);
+  addNode(AbbrevOffsetsRef);
   for (auto CURef : CURefs)
     addNode(CURef);
   if (auto E = createPaddingRef(DwarfSections.DebugInfo))
     return E;
   return finalizeSection();
 }
 
-Expected<SmallVector<DebugInfoCURef>> MCCASBuilder::splitDebugInfoSection() {
+Error MCCASBuilder::createDebugAbbrevSection(
+    ArrayRef<DebugAbbrevRef> AbbrevRefs) {
+  startSection(DwarfSections.Abbrev);
+  for (auto AbbrevRef : AbbrevRefs)
+    addNode(AbbrevRef);
+  if (auto E = createPaddingRef(DwarfSections.Abbrev))
+    return E;
+  return finalizeSection();
+}
+
+Expected<SmallVector<DebugAbbrevRef>>
+MCCASBuilder::splitAbbrevSection(ArrayRef<size_t> AbbrevOffsetVec) {
+  // If we are here, the Abbrev section _must_ exist.
+  if (!DwarfSections.Abbrev)
+    return createStringError(inconvertibleErrorCode(), "Invalid Dwarf");
+
+  const MCSection::FragmentListType &FragmentList =
+      DwarfSections.Abbrev->getFragmentList();
+
+  Expected<SmallVector<char, 0>> FullAbbrevData =
+      mergeMCFragmentContents(FragmentList);
+  if (!FullAbbrevData)
+    return FullAbbrevData.takeError();
+
+  // Sorted container of offsets such that, for any iterator `it`, the interval
+  // [*it, *(it+1)] describes where to find one Abbreviation contribution.
+  SmallSet<size_t, 8> AbbrevOffsets;
+  AbbrevOffsets.insert(AbbrevOffsetVec.begin(), AbbrevOffsetVec.end());
+  AbbrevOffsets.insert(FullAbbrevData->size());
+
+  size_t StartOffset = *AbbrevOffsets.begin();
+  if (StartOffset != 0)
+    return createStringError(inconvertibleErrorCode(), "Invalid Dwarf");
```

Comment:
is there a more targeted error message we could use here? "Expected nonzero start offset"?